lib/+mwbs/+RobotSensors/+IMUsensor/IMUsensorProc.m

```diff
@@ -0,0 +1,74 @@
+classdef IMUsensorProc < matlab.System
+    % IMUsensorProc This block computes the typical outputs of an IMU from the base pose and joints kinematics.
+    % 
+    % Assuming DoFs is the number of internal degrees of freedom of the robot:
+    % 
+    % Input:
+    %   - w_H_b: Base pose, 4x4 matrix representing the homogenous transformation between the the
+    %   base frame and the world frame.
+    %   - nu: Base and joints velocities: Vector of size 6 representing the linear and angular
+    %   velocity of the base frame, concatenated with a vector of size DoFs, representing the
+    %   velocity of the joints.
+    %   - nuDot: Base and joints accelerations: Vector of size 6 representing the linear and angular
+    %   acceleration of the base frame, concatenated with a vector of size DoFs, representing the
+    %   acceleration of the joints.
+    %   - Jimu: Jacobian of the IMU frame.
+    %   - JimuDotNu: Bias linear acceleration of the IMU frame with respect to the inertial (world)
+    %   frame.
+    % 
+    % Output:
+    %   The sensor frame orientation, linear acceleration and angular velocity w.r.t. the inertial
+    %   (world) frame. Linear acceleration and angular velocity are expressed in the world frame.
+    %   - w_rollPitchYaw: roll, pitch, yaw Euler angles in rad.
+    %   - w_linAcc: Proper linear acceleration in m/s^2.
+    %   - w_omega: Angular velocity in rad/s.
+    %   - imuOut: A [12x1] 1-D vector concatenating the previous three outputs and a placeholder for
+    %   other measurements (e.g. magnetometer).
+    %   [roll,pitch,yaw,accx,accy,accz,omegax,omegay,omegaz,0,0,0]'.
+    %   
+    % Parameters:
+    %   - Frame name: the name of the IMU frame. It should be specified in the URDF model.
+
+    % Public, tunable properties
+    properties
+
+    end
+
+    properties(DiscreteState)
+
+    end
+
+    % Pre-computed constants
+    properties(Access = private)
+
+    end
+
+    methods(Access = protected)
+        function setupImpl(obj)
+            % Perform one-time calculations, such as computing constants
+        end
+
+        function [imuOut,w_omega,w_linAcc,w_rollPitchYaw] = stepImpl(obj,Jimu,JimuDotNu,w_H_imu,nu,nudot)
+            % Implement algorithm. Calculate y as a function of inputs and discrete states.
+            
+            % Gyroscope measurements
+            w_imuVel = Jimu*nu;
+            w_omega = w_imuVel(4:6);
```

Comment:
This is missing a rotation ${}^S R_A$, where $S$ is sensor frame and $A$ is the world/absolute frame. See equation 4.3 in https://traversaro.github.io/traversaro-phd-thesis/traversaro-phd-thesis.pdf .

Reply:
yes Silvio, you're right. I was going to fix this later on after finishing https://github.com/robotology/whole-body-controllers/pull/122, and after checking those sections in your thesis about the sensor acceleration, just to be sure I wouldn't miss anything. That's why I specified in the issue and block documentation that the quantities where expressed in the **world frame**. But you're right, better do be accurate right away.

Reply:
If you like to merge as it is for me it is ok, probably better to open an issue to eventually fix it, thanks1

Reply:
I'll push here. Just quick question @traversaro . The WBT Jacobian block output is actually the jacobian expressed in a mixed frame right? B[A], i.e. the frame with origin of B and orientation of A, where B is the specified body frame and A the inertial frame. Such that `J*nu` gives the equivalent of:
$$
{}^{A}R _{F_i[A]} J_i \nu
$$

Where $A$ is the inertial frame and $J_i$ is the Jacobian of body frame $F_i$ expressed in body coordinates as described in Featherstone RBDA textbook.

Reply:
It should be the case, as in [KinDynComputations.cpp](https://github.com/robotology/idyntree/blob/master/src/high-level/src/KinDynComputations.cpp), [`m_frameVelRepr` is set to `MIXED_REPRESENTATION`](https://github.com/robotology/idyntree/blob/0c084c1717f71c8e507f9d9c8bad8f3b4919bdaf/src/high-level/src/KinDynComputations.cpp#L190).

Running the following in MATLAB command line [getFrameVelocityRepresentation()](https://github.com/robotology/idyntree/blob/0c084c1717f71c8e507f9d9c8bad8f3b4919bdaf/src/high-level/src/KinDynComputations.cpp#L478):
```
>> kinDyn = iDynTree.KinDynComputations()
>> kinDyn.getFrameVelocityRepresentation
ans =
  int64
   2
```
returns the default value: `2` => `MIXED_REPRESENTATION`. We get the same value while running `RobotVisualizer.prepareRobot()` which loads the robot model and prepares the robot visualization through the iDynTree bindings. Assuming that the Frame Velocity Representation is not explicitely changed later on through [`KinDynComputations::setFrameVelocityRepresentation`](https://github.com/robotology/idyntree/blob/0c084c1717f71c8e507f9d9c8bad8f3b4919bdaf/src/high-level/src/KinDynComputations.cpp#L483) method, we can assume that the **Jacobian** WBT block uses that type of representation. I've opened an issue to cross check that: #42 .

Reply:
> I'll push here. Just quick question @traversaro . The WBT Jacobian block output is actually the jacobian expressed in a mixed frame right? B[A], i.e. the frame with origin of B and orientation of A, where B is the specified body frame and A the inertial frame. Such that `J*nu` gives:
$$
\begin{equation}
{}^{F _i[A]}\mathrm{v} _{A,F _i} = {}^{F _i[A]}R _{F _i} J _i \nu
\end{equation}
$$

> Where $A$ is the inertial frame and $J_i$ is the Jacobian of body frame $F_i$ expressed in body coordinates as described in Featherstone RBDA textbook.

In this formula, what is ${}^{F _i[A]}R _{F _i}$ ? Anyway, for iDynTree convention that formula is incomplete because if you pass from the mixed jacobian to the body-fixed jacobian, also the convention used by floating base velocity in the $\nu$ vector changes, so you need to also account for that. See equation 3.70 of https://traversaro.github.io/traversaro-phd-thesis/traversaro-phd-thesis.pdf for the correct formula.


In any case, yes all WB-Toolbox bocks use `MIXED_REPRESENTATION` and i do not think there is any way to change it, so it is safe to assume that the jacobian is the mixed one according to iDynTree convention.

Reply:
sorry, I meant actually ${}^{F _i[A]}X _{F _i}$ instead of $R$. And the Jacobian was already left trivialized. Nevertheless I shouldn't mention Featherstone as he addresses the fixed base casein the sections I was thinking about. Anyway, it's much clearer now for me, after reading again the respective section in your thesis. Thanks!